File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
class Solution {
   public:
    int characterReplacement(string s, int k) {
        int st = 0;
        int en = 0;
        int n = s.size();
        int maxi_freq = INT_MIN;
        unordered_map<char, int> mp;
        int ans = 0;

        while (en < n) {
            mp[s[en]]++;

            maxi_freq = max(maxi_freq, mp[s[en]]);

            if ((en - st + 1) - maxi_freq <= k) {
                  ans = max(ans, en - st + 1);
            }else{
                mp[s[st]]--;
                st++;
            }
            
          
            en++;

        }
        return ans;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int max_depth = 0;
    int solve(TreeNode root){

        if(root == null)return 0;
        int ans = 0;
        int l = 0 , r = 0;
        if(root.left != null)
         l = Math.max(ans , 1 + solve(root.left));
        if(root.right != null)
         r = Math.max(ans , 1 + solve(root.right));
        max_depth = Math.max(l + r , max_depth);
        return Math.max(l , r);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        
        solve(root);
        if(root.left != null)
        diameterOfBinaryTree(root.left);
        if(root.right != null)
        diameterOfBinaryTree(root.right);

        return max_depth;
    }
}
